### PR TITLE
Fixes issue with dangling index being deleted instead of re-imported

### PR DESCRIPTION
**Pull request by abeyad**
Fixes an issue where a node that receives a cluster state
update with a brand new cluster UUID but without an
initial persistence block could cause indices to be wiped out,
preventing them from being reimported as dangling indices.
This commit only removes the in-memory data structures and
thus, are subsequently reimported as dangling indices.

### CODE REVIEW

**Comment by abeyad** [comment]
@ywelsch FYI

File: core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java [hunk 1]
```diff
@@ -199,10 +199,12 @@ public boolean nodesChanged() {
...
     }
 
-    // Determines whether or not the current cluster state represents an entirely
-    // different cluster from the previous cluster state, which will happen when a
-    // master node is elected that has never been part of the cluster before.
-    private boolean isNewCluster() {
+    /**
+     * Determines whether or not the current cluster state represents an entirely
+     * different cluster from the previous cluster state, which will happen when a
+     * master node is elected that has never been part of the cluster before.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java]: doesn't this also happen when the node first joins the cluster? if fos we should doc it.

Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java]: yes that's correct, i'll add it to the documentation

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java [hunk 2]
```diff
@@ -286,19 +286,6 @@ protected void doRun() throws Exception {
...
-            Index index = indexService.index();
-            IndexMetaData indexMetaData = event.state().metaData().index(index);
-            if (indexMetaData == null) {
-                assert false : "index" + index + " exists locally, doesn't have a metadata but is not part"
-                    + " of the delete index list. \nprevious state: " + event.previousState().prettyPrint()
-                    + "\n current state:\n" + event.state().prettyPrint();
-                logger.warn("[{}] isn't part of metadata but is part of in memory structures. removing", index);
-                indicesService.deleteIndex(index, "isn't part of metadata (explicit check)");
-            }
-        }
```
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: @ywelsch @bleskes I did some more investigation and I believe this block of code is unnecessary and hence removed it. This is because right after the `deleteIndices`, we call `removeUnallocatedIndices` which does the same thing as here - it removes the in-memory structures of indices no longer assigned to the node in the cluster state, and on a brand new cluster state, this will be the case.  The tests I wrote pass for this.

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: I wonder if we can make this behavior more prominent in the code, e.g. by changing the line
`if (state.blocks().disableStatePersistence()) {` in `IndicesClusterStateService.clusterChanged` to 
`if (state.blocks().disableStatePersistence() || event.isNewCluster()) {`.

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 3]
```diff
@@ -97,6 +100,49 @@ public void testRandomClusterStateUpdates() {
...
+        for (DiscoveryNode node : newClusterState.nodes()) {
+            IndicesClusterStateService indicesClusterStateService = clusterStateServiceMap.get(node);
+            ClusterState localState = adaptClusterStateToLocalNode(newClusterState, node);
+            ClusterState previousLocalState = adaptClusterStateToLocalNode(state, node);
+            indicesClusterStateService.clusterChanged(
+                new ClusterChangedEvent("simulated change 2", localState, previousLocalState));
+
+            // check that in memory data structures have been removed with the new cluster
+            for (ObjectCursor<IndexMetaData> indexMetaData : state.metaData().getIndices().values()) {
+                assertNull(indicesClusterStateService.indicesService.indexService(indexMetaData.value.getIndex()));
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: might be interesting to check that the index has been removed and not deleted?

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 4]
```diff
@@ -97,6 +100,49 @@ public void testRandomClusterStateUpdates() {
...
+        ClusterState newClusterState = ClusterState.builder(previousState)
+                                           .metaData(MetaData.builder(previousState.metaData()).clusterUUID(UUIDs.randomBase64UUID()))
+                                           .build();
+
+        // add index to the cluster
+        String name = "index_" + randomAsciiOfLength(8).toLowerCase(Locale.ROOT);
+        CreateIndexRequest request = new CreateIndexRequest(name, Settings.builder()
+                                                                      .put(SETTING_NUMBER_OF_SHARDS, randomIntBetween(1, 3))
+                                                                      .put(SETTING_NUMBER_OF_REPLICAS, 1)
+                                                                      .build()).waitForActiveShards(ActiveShardCount.ALL);
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: does this work? I thought that waitForActiveShards would not work with the mock ClusterChanges class?

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java (hunk 2 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: by removing this code altogether we also lose the assertion which informs us about new unexpected cases where in-memory index is not deleted by ClusterChangedEvent.indicesDeleted().

Comment (abeyad) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: The `state.blocks().disableStatePersistence()` in `IndicesClusterStateService.clusterChanged` removes the indices and then returns early, not processing anything else.  I am not sure that is what we want in the case of having a new cluster, but having missed the earlier cluster state updates, we are now processing this cluster state update with the new cluster UUID. It seems to me that we still want to process the cluster state update through the various methods (create indices, delete indices, etc). 

I do agree that we lose the assertion regarding this case in specific of a new cluster.  I'm not sure that this alone is worth keeping the code for. We could keep the loop and the assertion, and let the removal happen in `removeUnallocatedIndices`?  Or add a comment to `removeUnallocatedIndices` as well as an assertion such as:

```
private void removeUnallocatedIndices(final ClusterChangeEvent event) {
        // same as before

        for (AllocatedIndex<? extends Shard> indexService : indicesService) {
            Index index = indexService.index();
            if (indicesWithShards.contains(index) == false) {
                // if the index does not exist in the cluster state, then it should have been deleted earlier, unless its a brand new cluster (new cluster UUID)
                assert event.state.metaData().index(index) != null || event.isNewCluster(); 
                logger.debug("{} removing index, no shards allocated", index);
                indicesService.removeIndex(index, "removing index (no shards allocated)");
            }
        }
    }

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java (hunk 4 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: Yeah its useless, i'll remove it, I was experimenting and forgot to remove it

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java (hunk 3 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: Good point, I'll add that check

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java (hunk 2 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: correct, we should not return early when cluster uuid changes. Adding the assertion `assert event.state.metaData().index(index) != null || event.isNewCluster();` to `removeUnallocatedIndices` sounds great as a solution.

**Comment by abeyad** [comment]
@ywelsch I pushed https://github.com/elastic/elasticsearch/pull/19666/commits/6d45c822a4e91f19f098362f4e90a6f68c872a9c

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java [hunk 5]
```diff
@@ -322,6 +310,9 @@ private void removeUnallocatedIndices(final ClusterState state) {
         for (AllocatedIndex<? extends Shard> indexService : indicesService) {
             Index index = indexService.index();
             if (indicesWithShards.contains(index) == false) {
+                assert state.metaData().index(index) != null || event.isNewCluster() :
+                    "if the index does not exist in the cluster state, it should either " +
+                    "have been deleted or the cluster must be new";
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: can you explain in a comment here why it's ok to remove the index instead of deleting it when `event.isNewCluster()` (i.e. dangling indices)?
Can you also add `index` to the assertion error message.

File: core/src/test/java/org/elasticsearch/indices/cluster/AbstractIndicesClusterStateServiceTestCase.java [hunk 6]
```diff
@@ -242,7 +242,7 @@ private boolean hasIndex(Index index) {
 
         private final IndexSettings indexSettings;
 
-        public MockIndexService(IndexSettings indexSettings) {
+        protected MockIndexService(IndexSettings indexSettings) {
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/AbstractIndicesClusterStateServiceTestCase.java]: unrelated change -> please undo

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 7]
```diff
@@ -267,7 +318,7 @@ private IndicesClusterStateService createIndicesClusterStateService() {
         final ThreadPool threadPool = mock(ThreadPool.class);
         final Executor executor = mock(Executor.class);
         when(threadPool.generic()).thenReturn(executor);
-        final MockIndicesService indicesService = new MockIndicesService();
+        final MockIndicesService indicesService = new RecordingIndicesService();
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: maybe make this a parameter on `randomInitialClusterState` (i.e. `Supplier<MockIndicesService>`)

Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: or just leave it up to the caller to pass the `AllocatedIndices<? extends Shard, ? extends AllocatedIndex<? extends Shard>> indicesService` in.

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 8]
```diff
@@ -97,6 +101,53 @@ public void testRandomClusterStateUpdates() {
         logger.info("Final cluster state: {}", state.prettyPrint());
     }
 
+    public void testSimulateMasterDataDirectoryWiped() {
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: I feel that the test method name does not really reflect what is happening in the test. Should it be something like "testJoiningNewClusterRemovesExistingInMemoryIndexStructures"?
Feel free to add some comments as well to what is actually tested here.

Last, I think that we don't need to simulate a whole cluster to test the behavior here. Checking one single node is enough.

**Comment by abeyad** [comment]
@ywelsch I pushed https://github.com/elastic/elasticsearch/pull/19666/commits/23c1dede7220f89b06d2f9f681ba1cc295795ee5 that addresses your feedback

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 9]
```diff
@@ -97,7 +102,71 @@ public void testRandomClusterStateUpdates() {
...
+                                                                      .put(SETTING_NUMBER_OF_SHARDS, randomIntBetween(1, 3))
+                                                                      .put(SETTING_NUMBER_OF_REPLICAS, 1)
+                                                                      .build());
+        ClusterState state = cluster.createIndex(previousState, request);
+        assertTrue(state.metaData().hasIndex(name));
+
+        // pick a data node to simulate the cluster state change on
+        DiscoveryNode node = null;
+        for (DiscoveryNode discoveryNode : clusterStateServiceMap.keySet()) {
+            if (discoveryNode.isDataNode()) {
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: maybe you should pick a node that has actually a shard of the index assigned to it. This allows to check that the node actually had in-memory structure of the index before it was removed.
```
DiscoveryNode nodeWithShard = state.nodes().get(randomFrom(state.routingTable().index(name).shardsWithState(INITIALIZING)).currentNodeId());
```

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 10]
```diff
@@ -97,7 +102,71 @@ public void testRandomClusterStateUpdates() {
...
+        IndicesClusterStateService indicesClusterStateService = clusterStateServiceMap.get(node);
+        ClusterState localState = adaptClusterStateToLocalNode(state, node);
+        ClusterState previousLocalState = adaptClusterStateToLocalNode(previousState, node);
+        indicesClusterStateService.clusterChanged(
+            new ClusterChangedEvent("simulated change 1", localState, previousLocalState));
+
+        // check that cluster state has been properly applied to node
+        assertClusterStateMatchesNodeState(localState, indicesClusterStateService);
+
+        indicesClusterStateService = clusterStateServiceMap.get(node);
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: indicesClusterStateService should not have changed, so no need to refetch from the map.

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 11]
```diff
@@ -97,7 +102,71 @@ public void testRandomClusterStateUpdates() {
...
+                node = discoveryNode;
+                break;
+            }
+        }
+        assertNotNull(node);
+        IndicesClusterStateService indicesClusterStateService = clusterStateServiceMap.get(node);
+        ClusterState localState = adaptClusterStateToLocalNode(state, node);
+        ClusterState previousLocalState = adaptClusterStateToLocalNode(previousState, node);
+        indicesClusterStateService.clusterChanged(
+            new ClusterChangedEvent("simulated change 1", localState, previousLocalState));
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: instead of "simulated change 1" can you provide a text that describes the actual change? e.g. "index created" or "cluster uuid changed".

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 12]
```diff
@@ -97,7 +102,71 @@ public void testRandomClusterStateUpdates() {
...
+        indicesClusterStateService = clusterStateServiceMap.get(node);
+        localState = adaptClusterStateToLocalNode(newClusterState, node);
+        previousLocalState = adaptClusterStateToLocalNode(state, node);
+        indicesClusterStateService.clusterChanged(
+            new ClusterChangedEvent("simulated change 2", localState, previousLocalState));
+
+        // check that in memory data structures have been removed with the new cluster,
+        // but the persistent data is still there
+        RecordingIndicesService indicesService = (RecordingIndicesService) indicesClusterStateService.indicesService;
+        for (ObjectCursor<IndexMetaData> indexMetaData : state.metaData().getIndices().values()) {
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: just `for (IndexMetaData indexMetaData : state.metaData())`

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 13]
```diff
@@ -97,7 +102,71 @@ public void testRandomClusterStateUpdates() {
...
+     * structures to be removed (see {@link IndicesClusterStateService#clusterChanged(ClusterChangedEvent)}),
+     * but in the case where the node joined and was a few cluster state updates behind, it would
+     * not have received the cluster block, in which case we still need to remove the in-memory
+     * structures while ensuring the data remains on disk.  This test executes this particular
+     * scenario.
+     */
+    public void testJoiningNewClusterOnlyRemovesInMemoryIndexStructures() {
+        // we have an IndicesClusterStateService per node in the cluster
+        final Map<DiscoveryNode, IndicesClusterStateService> clusterStateServiceMap = new HashMap<>();
+        ClusterState previousState = randomInitialClusterState(clusterStateServiceMap, RecordingIndicesService::new);
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: If we separate `updateNodes(state, clusterStateServiceMap);` out of `randomInitialClusterState`, there is no need to have a `clusterStateServiceMap` in this test.

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 14]
```diff
@@ -97,7 +102,71 @@ public void testRandomClusterStateUpdates() {
...
+        ClusterState newClusterState = ClusterState.builder(previousState)
+                                           .metaData(MetaData.builder(previousState.metaData()).clusterUUID(UUIDs.randomBase64UUID()))
+                                           .build();
+
+        // add index to the cluster
+        String name = "index_" + randomAsciiOfLength(8).toLowerCase(Locale.ROOT);
+        CreateIndexRequest request = new CreateIndexRequest(name, Settings.builder()
+                                                                      .put(SETTING_NUMBER_OF_SHARDS, randomIntBetween(1, 3))
+                                                                      .put(SETTING_NUMBER_OF_REPLICAS, 1)
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: isn't this the default? just omit in that case.

**Comment by abeyad** [comment]
@ywelsch I pushed https://github.com/elastic/elasticsearch/pull/19666/commits/1588d8741053f933a17c133320b76b847cbdcb0b
I think (hope) you will like the organization of the test better here.

**Comment by ywelsch** [comment]
LGTM. Thanks @abeyad!

**Comment by abeyad** [comment]
Thanks @ywelsch !
